interpreter/cling/lib/Interpreter/DefinitionShadower.cpp

```diff
@@ -132,6 +140,18 @@ namespace cling {
     // : reference to 'xxx' is ambiguous" in `class C {}; class C; C foo;`.
     if (!Previous.empty() && !isDefinition(D))
       D->setInvalidDecl();
+
+    // Diagnose shadowing of decls in the `std` namespace (see ROOT-5971).
+    // For unnamed macros, the input is ingested in a single `Interpreter::process()`
+    // call. Do not emit the warning in that case, as all references are local
+    // to the wrapper function and this diagnostic might be misleading.
+    if (shadowsDeclInStd
+        && ((m_Interp.getInputFlags() & (Interpreter::kInputFromFile
+                                         | Interpreter::kIFFLineByLine))
+             != Interpreter::kInputFromFile)) {
```

Comment:
This means we also warn for `m_Interp.getInputFlags() == 0`, right? Does that mean all of REPL and things like cppyy and RDF?

Reply:
Yes, that is right.  It means that we warn if definition shadowing did its job (that typically implies that we entered via `Interpreter::process()`), and that the user was trying to shadow a global declaration that conflicts with something we pulled in to the global scope from `std` (via the implicit `using namespace std;`).

In other words, this happens for REPL, but not for cppyy/RDF as they generate wrapper functions that should not be touched by DefinitionShadower. See below for an example.

```c++
root [0] unsigned data = 0;
ROOT_prompt_0:1:1: warning: 'data' shadows a declaration with the same name in the 'std' namespace; consider using '::data' to reference this declaration
unsigned data = 0;
^
root [1] data
ROOT_prompt_1:1:1: error: reference to 'data' is ambiguous
data
^
ROOT_prompt_0:1:10: note: candidate found by name lookup is '__cling_N50::data'
unsigned data = 0;
         ^
/usr/lib/gcc/x86_64-pc-linux-gnu/11.1.0/../../../../include/c++/11.1.0/bits/range_access.h:319:5: note: candidate found by name lookup is 'std::data'
    data(initializer_list<_Tp> __il) noexcept
    ^
/usr/lib/gcc/x86_64-pc-linux-gnu/11.1.0/../../../../include/c++/11.1.0/bits/range_access.h:290:5: note: candidate found by name lookup is 'std::data'
    data(_Container& __cont) noexcept(noexcept(__cont.data()))
    ^
/usr/lib/gcc/x86_64-pc-linux-gnu/11.1.0/../../../../include/c++/11.1.0/bits/range_access.h:300:5: note: candidate found by name lookup is 'std::data'
    data(const _Container& __cont) noexcept(noexcept(__cont.data()))
    ^
/usr/lib/gcc/x86_64-pc-linux-gnu/11.1.0/../../../../include/c++/11.1.0/bits/range_access.h:310:5: note: candidate found by name lookup is 'std::data'
    data(_Tp (&__array)[_Nm]) noexcept
    ^
root [2] ::data
(unsigned int) 0
```